pkg/event/target/queuestore.go

```diff
@@ -0,0 +1,219 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package target
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/minio/minio/pkg/event"
+)
+
+const (
+	maxLimit = 10000 // Max store limit.
+	eventExt = ".event"
+)
+
+type QueueStore struct {
+	sync.RWMutex
+	directory string
+	eC        uint16
+	limit     uint16
+}
+
+// NewQueueStore - Filestore for persisting events.
+func NewQueueStore(directory string, limit uint16) *QueueStore {
+	if limit == 0 {
+		limit = maxLimit
+	}
+	queueStore := &QueueStore{
+		directory: directory,
+		limit:     limit,
+	}
+	return queueStore
+}
+
+// Staging directory.
+func tmpPath(directory string) string {
+	return directory + "/tmp"
+}
+
+// Open - Creates the directory if not present.
+func (store *QueueStore) Open() error {
+	store.Lock()
+	defer store.Unlock()
+
+	if err := store.resetTmp(); err != nil {
+		return err
+	}
+
+	tmpDir := tmpPath(store.directory)
+	if terr := os.MkdirAll(tmpDir, os.FileMode(0770)); terr != nil {
+		return terr
+	}
+
+	eCount := uint16(len(store.listAll()))
+	if eCount >= store.limit {
+		return ErrLimitExceeded
+	} else {
+		store.eC = eCount
+	}
+
+	return nil
+}
+
+// write - writes event to the directory.
+func (store *QueueStore) write(directory string, key string, e event.Event) error {
+
+	tp := filepath.Join(tmpPath(store.directory), key)
+
+	// Creates the temporary file.
+	temp, cerr := os.Create(tp)
+	if cerr != nil {
+		return cerr
+	}
+
+	// Marshalls the event.
+	eventData, err := json.Marshal(e)
+	if err != nil {
+		return err
+	}
+
+	// Writes to the temporary file.
+	if _, err := temp.Write(eventData); err != nil {
+		return err
+	}
+
+	p := filepath.Join(store.directory, key+eventExt)
+	// Rename the temp file to the destination file.
+	if err := os.Rename(tp, p); err != nil {
+		return err
+	}
+
+	// Increment the event count.
+	store.eC += 1
+
+	return nil
+}
+
+// Put - puts a event to the store.
+func (store *QueueStore) Put(e event.Event) error {
+	store.Lock()
+	defer store.Unlock()
+	if store.eC >= store.limit {
+		return ErrLimitExceeded
+	}
+	key := strconv.Itoa(time.Now().Nanosecond())
+	return store.write(store.directory, key, e)
+}
+
+// Get - gets a event from the store.
+func (store *QueueStore) Get(key string) (event.Event, error) {
+	store.RLock()
+	defer store.RUnlock()
+
+	var event event.Event
+
+	filepath := filepath.Join(store.directory, key)
+
+	if _, err := os.Stat(filepath); err != nil {
+		return event, err
+	}
+
+	// Open the file.
+	eFile, oerr := os.Open(filepath)
+	if oerr != nil {
+		return event, oerr
+	}
+
+	eventData := make([]byte, 1024*2)
```

Comment:
For reading as well just use `ioutil.ReadAll()`  no need to make sure we are only reading 2K worth of data, it may go beyond this value as well one can never tell. Also looks like `key+eventExt` is missing while reading.

Reply:
When we see a stale or zero byte file, we can simply ignore it and delete it.

Reply:
done